sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_cluster.asciidoc

```diff
@@ -0,0 +1,200 @@
+
+[%header,cols="10,5"]
+|===
+
+|`cluster_group<Dimensions> nd_item::get_cluster_group()`
+|Returns the consituent `cluster_group` in the kernel, representing this `cluster_group`'s overall position in the `nd_range`
+|===
+
+
+The `cluster_group` class to have the following methods, to access the various id's 
+
+
+[%header,cols="5,5"]
+|===
+|Method
+|Description
+
+|`size_t get_group_linear_id(void)`
+|Returns the flattened id of the calling `group` within the cluster
+
+|`size_t get_local_linear_id(void)`
+|Returns the flattened index of the calling `workitem` within the cluster
+
+|`size_t get_group_linear_range(void)`
+|Returns the toal number of `groups` within the cluster
+
+|`size_t get_group_id(size_t)`
+|Returns the id of the calling workgroup along the specified dimension
+
+|`size_t get_group_linear_range(size_t)`
+|Returns the toal number of `groups` within the cluster in the specified dimension
+
+|`size_t get_local_linear_range(void)`
+|Returns the total number of `workItems` within the cluster
+|===
+
+
+To obain the total number of clusters within the group, and getting the cluster id, `nd_item` to have 2 more class methods, namely `get_cluster_range` and `get_cluster_id`
+
+[%header,cols="10,5"]
+|===
+
+|`size_t nd_item::get_cluster_range(size_t)`
+|Returns the total number of `cluster_group`s in the kernel across the specified dimension
+
+|`size_t nd_item::get_cluster_id(size_t)`
+|Returns the id of the cluster across the specified dimension
+|===
+
+
+CC 9.0 devices feature the ability to synchronize all the workgroups within the cluster, and to achieve that in SYCL, this proposes add an overload to the existing `sycl::group_barrier` function, accepting `cluster_group` type.
+
+[%header,cols="10,5"]
+|===
+
+|`void sycl::group_barrier(cluster_group<Dimensions> G)`
+|Synchronizes all workgroup within the cluster
+|===
+
+To prevent un-defined behaviour, in the case where these range's or id's are queried without the cluster dimensions mentioned at launch can be considered as a cluster launch with cluster dimension as 1.
+
+=== Accessing another workgroup's local memory
+
+Workgroup's within the cluster have the ability to access another workgroup's local memory. The local memory addresses can only be addressed by the `workItems` of that workgroup. Therefore, to access another workgroup's local memory, the address would need to be mapped such that it is addressable by calling `workItem` belonging to another workgroup. A public member function of the `cluster_group` class, `get_cluster_local_pointer` will perform the above and return the pointer which can then be dereferenced by the calling `workItem`
+
+
+[%header,cols="10,5"]
+|===
+
+|`T*  get_cluster_local_pointer(T* addr, size_t group_id)`
```

Comment:
I would maybe suggest naming this something like `map_cluster_local_pointer` instead since it's doing more than just getting a pointer.

Reply:
I think it would make sense for this interface to use `multi_ptr` to constrain it to pointers to local memory, and avoid UB. @Naghasan what do think think?

Reply:
What are the control-flow semantics of the equivalant feature in CUDA, does this need to be called in convergent control flow, and must all work-items in the work-group take the same parameters?

Also is this function synchronous, when it returns is the pointer returned immediately accessible in the current work-group?

Reply:
> Each thread can call this independently. The only requirement is that the thread block of which the shared memory is being accessed, should have started executing and not finished execution, which can be guaranteed by a cluster wide synchronization. (source Cuda programming guide)

So typically group-based functions in SYCL are convergent, in that all work-items in a particular scope must call the function in the same control flow, however, I believe in order to use this feature effectively it must be called per-work-item, and will only work on hardware with independent forward progress of work-items.

In the future if we come to generalise this extension across all backends I think we will need to address this, but since this is CUDA-specific I think it is fine for now. Though we should explain these semantics in definition of the function.

> The scope is per-thread, so when this pointer is returned it is accessible to that thread only

We should explain this in the definition of the function as well so it's clear that the mapping is blocking and will be complete by the time the function returns.

Reply:
> So typically group-based functions in SYCL are convergent, in that all work-items in a particular scope must call the function in the same control flow

But isn't that only for group algorithms which collaboratively compute / share data. I believe in this case there is no "collaboration" so to speak.

The only requirement to access another work-group's local memory, is that the work-group should exist, implying that local memory exists. Then Any work-item can map that local memory start address and start dereferencing it as required. 

> and will only work on hardware with independent forward progress of work-items.

Assume a device with the lowest level of independent forward progress at a sub-group level. On that device, even if only say 1 work-item calls this function, would this be conceptually different than - 
```cpp
if (sub_group.get_local_linear_id() == 0) {
    any_function_foo();
}
```
I say so because note, the work-item calling `map_cluster_local_pointer` does not **directly** depend on another work-item directly within it's group, as in any another work-item is not responsible to return the address, but rather the existence of another work-group within the cluster-group.

Reply:
> I say so because , the work-item calling map_cluster_local_pointer does not directly depend on another work-item within the group, as in any another work-item is not responsible to return the address, but rather the existence of another work-group within the cluster-group.

BTW, this is indeed very Nvidia specific, but it's a cuda only extension so it's fine for now. If another hardware starts to support this feature, and has another set of requirement, they would need to be incorporated somehow, depending on what they are.

But, one of my main discussions above is that maybe independent forward progress of work-items is not required for this feature from what I gather, seeing how it is exposed at the PTX level

Reply:
>But isn't that only for group algorithms which collaboratively compute / share data. I believe in this case there is no "collaboration" so to speak.

That's right, this function I think we need to make. a clear distinction distinction between this kind of function and group functions/algorithms, even though these involve a group id.

> I say so because , the work-item calling this function does not directly depend on another work-item within the group, as in any another work-item is not responsible to return the address, but rather the existence of another work-group within the cluster-group.

So this is right, `map_cluster_local_pointer` can execute independently within the work-item with no dependency on other work-items in the group, however, depending on how the function lowers to instructions and whether the hardware supports independent forward progress of work-items, it may or may not be able to execute in separate work-items with different parameters concurrently.

But I think we can define it as a built-in function which executes within a work-item, which does not require convergence, permitting different parameters in each work-item, but leave it unspecified whether these are then executed concurrently or masked and executed in sequence. Though this is perhaps thinking too far in the future, as this is still a CUDA-specific extension.